### PR TITLE
feat(api): support metric API Prometheus format & add statistic metric api 

### LINKED ISSUE #1656: [Question] 有提供metrics监控接口吗？能否通过prometheus的json_exporter实现监控指标数据的转换
### Problem Type (问题类型)

others (please edit later)

### Before submit

- [X] 我已经确认现有的 [Issues](https://github.com/hugegraph/hugegraph/issues) 与 [FAQ](https://hugegraph.github.io/hugegraph-doc/guides/faq.html) 中没有相同 / 重复问题

### Environment (环境信息)

- Server Version: v0.11.2
- Backend: RocksDB x nodes, HDD or SSD 

### Your Question (问题描述)

在prometheus的环境下，访问ip:18080/metrics ，得到的页面结果如下：
d9536287313c814f99335d29b2460c1
9e5b3b1cdaf38e1b1a451cec022adc5

出现这个错误的原因是因为prometheus无法转化metrics的数据导致的吗？目前有没有提供exporter来解决数据转换的问题呢？

### Vertex/Edge example (问题点 / 边数据举例)

_No response_

### Schema [VertexLabel, EdgeLabel, IndexLabel] (元数据结构)

_No response_

### PR DESCRIPTION
**Pull request by SunnyBoy-WYH**
fix #1656 #1356  (will enhance it later)

术语：
统计指标：指【请求总数、成功数、失败数、平均响应时间、最大响应时间】五个指标

基础指标：指serverReporter中原有的gauges、counters、histograms等

思路：
image

1.增加原有Metric接口的prom格式返回：

> \# HELP hugegraph_info
\# TYPE hugegraph_info untyped
hugegraph_info{version="0.69",} 1.0
\# HELP org_apache_hugegraph_backend_cache_Cache_edge_hugegraph_capacity
\# TYPE org_apache_hugegraph_backend_cache_Cache_edge_hugegraph_capacity gauge
org_apache_hugegraph_backend_cache_Cache_edge_hugegraph_capacity 1000000
\# HELP org_apache_hugegraph_backend_cache_Cache_edge_hugegraph_expire
\# TYPE org_apache_hugegraph_backend_cache_Cache_edge_hugegraph_expire gauge
org_apache_hugegraph_backend_cache_Cache_edge_hugegraph_expire 600000
\# HELP org_apache_hugegraph_backend_cache_Cache_edge_hugegraph_hits

2.增加statistics接口，返回【请求总数、成功数、失败数、平均响应时间、最大响应时间】五个指标，同时支持prop格式返回
json：
> {
    "metrics/GET":{
        "FAILED_REQUEST":0,
        "MEAN_RESPONSE_TIME":10.047648352712768,
        "MAX_RESPONSE_TIME":27,
        "SUCCESS_REQUEST":12,
        "TOTAL_REQUEST":12
    },
    "metrics/statistics/GET":{
        "FAILED_REQUEST":0,
        "MEAN_RESPONSE_TIME":0.6234681965773761,
        "MAX_RESPONSE_TIME":2,
        "SUCCESS_REQUEST":3,
        "TOTAL_REQUEST":3
    }
}
prom格式：

> \# HELP hugegraph_info
\# TYPE hugegraph_info untyped
hugegraph_info{version="0.69",} 1.0
\# HELP metrics_GET
\# TYPE metrics_GET gauge
metrics_GET{name=FAILED_REQUEST,} 0
metrics_GET{name=MEAN_RESPONSE_TIME,} 10.047648352712768
metrics_GET{name=MAX_RESPONSE_TIME,} 27
metrics_GET{name=SUCCESS_REQUEST,} 12
metrics_GET{name=TOTAL_REQUEST,} 12

3.需要讨论一下 statistics接口的数据是否要合并在Metric接口中，目前没有合并是因为，原有指标结果JSON中只包含counter、summary等五种标准指标 ；

另外分开还有个好处是，如果后续想要按照图、接口查询统计指标，比较方便定制

### CODE REVIEW

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java [hunk 1]
```diff
@@ -22,55 +22,53 @@
...
+            MetricsUtil.registerMeter(API.class, "commit-succeed");
     private static final Meter ILLEGAL_ARG_ERROR_METER =
-                         MetricsUtil.registerMeter(API.class, "illegal-arg");
+            MetricsUtil.registerMeter(API.class, "illegal-arg");
     private static final Meter EXPECTED_ERROR_METER =
-                         MetricsUtil.registerMeter(API.class, "expected-error");
+            MetricsUtil.registerMeter(API.class, "expected-error");
     private static final Meter UNKNOWN_ERROR_METER =
-                         MetricsUtil.registerMeter(API.class, "unknown-error");
+            MetricsUtil.registerMeter(API.class, "unknown-error");
```
Comment (JackyYangPassion) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java:65]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java [hunk 2]
```diff
@@ -22,55 +22,53 @@
...
 
     public static HugeGraph graph(GraphManager manager, String graph) {
         HugeGraph g = manager.graph(graph);
         if (g == null) {
             throw new NotFoundException(String.format(
-                      "Graph '%s' does not exist",  graph));
+                    "Graph '%s' does not exist", graph));
```
Comment (JackyYangPassion) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java:71]: prefer to keep the origin alignment

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java [hunk 3]
```diff
@@ -97,7 +95,7 @@ public static <R> R commit(HugeGraph g, Callable<R> callable) {
             SUCCEED_METER.mark();
             return result;
         } catch (IllegalArgumentException | NotFoundException |
-                 ForbiddenException e) {
+                ForbiddenException e) {
```
Comment (JackyYangPassion) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java:98]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java [hunk 4]
```diff
@@ -141,7 +139,7 @@ protected static void checkUpdatingBody(Checkable body) {
     }
 
     protected static void checkCreatingBody(
-                          Collection<? extends Checkable> bodies) {
+            Collection<? extends Checkable> bodies) {
```
Comment (JackyYangPassion) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java:142]: ditto

File: hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java [hunk 5]
```diff
@@ -165,8 +190,8 @@ public void testMetricsBackend() {
                     String key = (String) e.getKey();
                     value = e.getValue();
                     Assert.assertTrue(String.format(
-                                      "Expect map value for key %s but got %s",
-                                      key, value),
+                                              "Expect map value for key %s but got %s",
```
Comment (JackyYangPassion) [inline hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java:193]: prefer to keep the origin alignment

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java [hunk 6]
```diff
@@ -22,55 +22,53 @@
...
-
+            .getSubtype();
     public static final String ACTION_APPEND = "append";
     public static final String ACTION_ELIMINATE = "eliminate";
     public static final String ACTION_CLEAR = "clear";
-
+    protected static final Logger LOG = Log.logger(API.class);
     private static final Meter SUCCEED_METER =
-                         MetricsUtil.registerMeter(API.class, "commit-succeed");
+            MetricsUtil.registerMeter(API.class, "commit-succeed");
```
[the remark this answers is not in the document]
Comment (SunnyBoy-WYH) [reply hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java:59]: okay,i used hg-code-style.xml with idea ,so it auto format when save.

**Comment by imbajin** [comment]
@SunnyBoy-WYH seems CI failed? could check the CI status in your **personal repo** (submit a PR to your repo, u could check/test CI there)

**Comment by SunnyBoy-WYH** [comment]
> @SunnyBoy-WYH seems CI failed? could check the CI status in your **personal repo** (submit a PR to your repo, u could check/test CI there)

Nice idea! ( when i push this pr, the local test passed, and a few days before i am looking for the ci problem,but ci info not enough.) i will try it again.

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    public void filter(ContainerRequestContext requestContext,
+                       ContainerResponseContext responseContext)
+            throws IOException {
+        // Grab corresponding request / response info from context;
+        String method = requestContext.getRequest().getMethod();
+        String path = requestContext.getUriInfo().getPath();
+        String metricsName = join(path, method);
+
+        MetricsUtil.registerCounter(
+                           join(metricsName, METRICS_PATH_TOTAL_COUNTER))
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java:59]: seems we can move to the previous line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * Use filter to log request info
+     *
+     * @param requestContext  requestContext
+     * @param responseContext responseContext
+     */
+    @Override
+    public void filter(ContainerRequestContext requestContext,
+                       ContainerResponseContext responseContext)
+            throws IOException {
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java:52]: prefer to align with ContainerResponseContext

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsKeys.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+package org.apache.hugegraph.metrics;
+
+public enum MetricsKeys {
+    // 最大响应时间 Maximum response time
+    // 平均响应时间 Mean response time
+    // 请求总数 Total Requests
+    // 失败数   Failed Requests
+    // 成功数   Success Requests
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsKeys.java:25]: remove unused comments or translate them?

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java [hunk 10]
```diff
@@ -24,12 +24,45 @@
...
 import com.codahale.metrics.Meter;
 import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Snapshot;
 import com.codahale.metrics.Timer;
 
 public class MetricsUtil {
 
-    private static final MetricRegistry REGISTRY =
-                                        MetricManager.INSTANCE.getRegistry();
+    public static final String METRICS_PATH_TOTAL_COUNTER = "TOTAL_COUNTER";
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:32]: does these name will be showed to users? if yes please rename to total_counter style?

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java [hunk 11]
```diff
@@ -40,15 +73,147 @@ public static Counter registerCounter(Class<?> clazz, String name) {
...
+
+    public static String replaceDotDashInKey(String orgKey) {
+        return orgKey.replace(".", "_").replace("-", "_");
+    }
+
+    public static String replaceSlashInKey(String orgKey) {
+        return orgKey.replace("/", "_");
+    }
+
+    public static void writePrometheus(StringBuilder promeMetrics,
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:105]: writePrometheusFormat

File: hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java [hunk 12]
```diff
@@ -17,20 +17,24 @@
...
 import org.junit.Test;
 
-import org.apache.hugegraph.testutil.Assert;
+import jakarta.ws.rs.core.Response;
 
 public class MetricsApiTest extends BaseApiTest {
 
-    private static String path = "/metrics";
+    private static final String path = "/metrics";
+    private static final String statistics_path = path + "/statistics";
```
Comment (javeme) [inline hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java:31]: statisticsPath

File: hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java [hunk 13]
```diff
@@ -17,20 +17,24 @@
...
 
     @Test
-    public void testMetricsAll() {
-        Response r = client().get(path);
+    public void testBaseMetricsAll() {
+        Map<String, Object> params = new HashMap<>();
+        params.put("type", "json");
+        Response r = client().get(path,params);
```
Comment (javeme) [inline hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java:37]: expect a space after the ","

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 14]
```diff
@@ -158,4 +177,253 @@ public String timers() {
...
+        Map<String, Map<String, Object>> metricMap = statistics();
+
+        if (type != null && type.equals(JSON_STR)) {
+            return JsonUtil.toJson(metricMap);
+        }
+        return statisticsProm(metricMap);
+
+    }
+
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:211]: remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 15]
```diff
@@ -158,4 +177,253 @@ public String timers() {
...
+        result.put("gauges", reporter.gauges());
+        result.put("counters", reporter.counters());
+        result.put("histograms", reporter.histograms());
+        result.put("meters", reporter.meters());
+        result.put("timers", reporter.timers());
+        return JsonUtil.toJson(result);
+    }
+
+    private String baseMetricPrometheusAll() {
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:224]: remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 16]
```diff
@@ -158,4 +177,253 @@ public String timers() {
...
     }
+
+
+    @GET
+    @Timed
+    @Produces(APPLICATION_TEXT_WITH_CHARSET)
+    @RolesAllowed({"admin", "$owner= $action=metrics_read"})
+    public String all(@Context GraphManager manager,
+                      @QueryParam("type") String type) {
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:188]: remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+@PreMatching
+public class PathFilter implements ContainerRequestFilter {
+
+
+    @Override
+    public void filter(ContainerRequestContext context)
+            throws IOException {
+
+        context.setProperty("RequestTime", System.currentTimeMillis());
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java:39]: remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+import jakarta.ws.rs.container.ContainerRequestFilter;
+import jakarta.ws.rs.container.PreMatching;
+import jakarta.ws.rs.ext.Provider;
+
+@Provider
+@Singleton
+@PreMatching
+public class PathFilter implements ContainerRequestFilter {
+
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java:33]: remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+            MetricsUtil.registerCounter(
+                               join(metricsName, METRICS_PATH_SUCCESS_COUNTER))
+                       .inc();
+        } else {
+            MetricsUtil.registerCounter(
+                               join(metricsName, METRICS_PATH_FAILED_COUNTER))
+                       .inc();
+        }
+
+        //  get responseTime
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java:73]: expect just one space after "//"

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+        String metricsName = join(path, method);
+
+        MetricsUtil.registerCounter(
+                           join(metricsName, METRICS_PATH_TOTAL_COUNTER))
+                   .inc();
+        if (responseContext.getStatus() == 200 ||
+            responseContext.getStatus() == 201 ||
+            responseContext.getStatus() == 202) {
+            MetricsUtil.registerCounter(
+                               join(metricsName, METRICS_PATH_SUCCESS_COUNTER))
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java:65]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+        String method = requestContext.getRequest().getMethod();
+        String path = requestContext.getUriInfo().getPath();
+        String metricsName = join(path, method);
+
+        MetricsUtil.registerCounter(
+                           join(metricsName, METRICS_PATH_TOTAL_COUNTER))
+                   .inc();
+        if (responseContext.getStatus() == 200 ||
+            responseContext.getStatus() == 201 ||
+            responseContext.getStatus() == 202) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java:63]: add a method `boolean statusOK(status)`?

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    @Override
+    public void filter(ContainerRequestContext context)
+            throws IOException {
+
+        context.setProperty("RequestTime", System.currentTimeMillis());
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java:38]: expect request_time style

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+@Singleton
+@PreMatching
+public class PathFilter implements ContainerRequestFilter {
+
+    @Override
+    public void filter(ContainerRequestContext context)
+            throws IOException {
+
+        context.setProperty("request_time", System.currentTimeMillis());
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java:38]: remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+@Provider
+@Singleton
+@PreMatching
+public class PathFilter implements ContainerRequestFilter {
+
+    @Override
+    public void filter(ContainerRequestContext context)
+            throws IOException {
+
+        context.setProperty("request_time", System.currentTimeMillis());
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/PathFilter.java:37]: can we define a const var for "request_time"?

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 25]
```diff
@@ -158,4 +177,253 @@ public String timers() {
...
+    @Produces(APPLICATION_TEXT_WITH_CHARSET)
+    @RolesAllowed({"admin", "$owner= $action=metrics_read"})
+    public String statistics(@QueryParam("type") String type) {
+        Map<String, Map<String, Object>> metricMap = statistics();
+
+        if (type != null && type.equals(JSON_STR)) {
+            return JsonUtil.toJson(metricMap);
+        }
+        return statisticsProm(metricMap);
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:208]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java (hunk 14 above)
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:211]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java (hunk 15 above)
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:224]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 26]
```diff
@@ -158,4 +177,253 @@ public String timers() {
...
+            }
+        }
+        return promMetric.toString();
+    }
+
+    private String joinWithSlash(String path1, String path2) {
+        return String.join("/", path1, path2);
+    }
+
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:428]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsKeys.java [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+
+    FAILED_REQUEST(4, "FAILED_REQUEST"),
+
+    SUCCESS_REQUEST(5, "SUCCESS_REQUEST"),
+    ;
+
+    private final byte code;
+    private final String name;
+
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsKeys.java:41]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java [hunk 28]
```diff
@@ -40,15 +73,146 @@ public static Counter registerCounter(Class<?> clazz, String name) {
...
+                            .append(FIFT_MIN_RATE_ATRR).append(timer.getFifteenMinuteRate())
+                            .append(END_LSTR);
+                promeMetrics.append(
+                        exportSnapshot(helpName, timer.getSnapshot()));
+            }
+        });
+    }
+
+    public static String exportSnapshot(final String helpName, final Snapshot snapshot) {
+        if (snapshot != null) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:191]: return first if if (snapshot == null)

File: hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java [hunk 29]
```diff
@@ -39,6 +43,27 @@ public void testMetricsAll() {
...
+        assertResponseStatus(200, r);
+    }
+
+    @Test
+    public void testStatisticsMetricsPromAll() {
+        Response r = client().get(statistics_path);
+        assertResponseStatus(200, r);
+    }
+
+
```
Comment (javeme) [inline hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java:66]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 30]
```diff
@@ -158,4 +177,247 @@ public String timers() {
...
+            }
+        }
+
+        MetricsUtil.writePrometheus(promMetric,
+                                    MetricManager.INSTANCE.getRegistry());
+
+        return promMetric.toString();
+    }
+
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:334]: remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsKeys.java [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package org.apache.hugegraph.metrics;
+
+public enum MetricsKeys {
+    // 最大响应时间 Maximum response time
+    // 平均响应时间 Mean response time
+    // 请求总数 Total Requests
+    // 失败数   Failed Requests
+    // 成功数   Success Requests
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsKeys.java:25]: remove unused comments or translate them?

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java (hunk 10 above)
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:32]: does these name will be showed to users? if yes please rename to total_counter style?

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java [hunk 32]
```diff
@@ -40,15 +73,145 @@ public static Counter registerCounter(Class<?> clazz, String name) {
...
+
     public static Histogram registerHistogram(Class<?> clazz, String name) {
         return REGISTRY.histogram(MetricRegistry.name(clazz, name));
     }
 
+    public static Histogram registerHistogram(String name) {
+        return REGISTRY.histogram(name);
+    }
+
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:88]: remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java [hunk 33]
```diff
@@ -40,15 +73,145 @@ public static Counter registerCounter(Class<?> clazz, String name) {
...
+    public static String replaceDotDashInKey(String orgKey) {
+        return orgKey.replace(".", "_").replace("-", "_");
+    }
+
+    public static String replaceSlashInKey(String orgKey) {
+        return orgKey.replace("/", "_");
+    }
+
+    public static void writePrometheus(StringBuilder promeMetrics,
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:105]: writePrometheusFormat

File: hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java (hunk 12 above)
Comment (javeme) [inline hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java:31]: rename the var name to statisticsPath style

File: hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java (hunk 13 above)
Comment (javeme) [inline hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java:37]: expect a space after the "path,"

**Review by javeme — COMMENTED** [review]
there are still some comments to be addressed, thanks

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 34]
```diff
@@ -325,13 +323,12 @@ private String baseMetricPrometheusAll() {
             }
         }
 
-        MetricsUtil.writePrometheus(promMetric,
+        MetricsUtil.writePrometheusFormat(promMetric,
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:326]: don't forget update the 327 alignment

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java [hunk 35]
```diff
@@ -102,7 +102,7 @@ public static String replaceSlashInKey(String orgKey) {
         return orgKey.replace("/", "_");
     }
 
-    public static void writePrometheus(StringBuilder promeMetrics,
+    public static void writePrometheusFormat(StringBuilder promeMetrics,
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:105]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java [hunk 36]
```diff
@@ -49,6 +49,8 @@ public class API {
     public static final String APPLICATION_JSON = MediaType.APPLICATION_JSON;
     public static final String APPLICATION_JSON_WITH_CHARSET =
                                APPLICATION_JSON + ";charset=" + CHARSET;
+    public static final String APPLICATION_TEXT_WITH_CHARSET =
+            MediaType.TEXT_PLAIN + ";charset=" + CHARSET;
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java:53]: also align with APPLICATION_TEXT_WITH_CHARSET

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java [hunk 37]
```diff
@@ -0,0 +1,83 @@
...
+import jakarta.ws.rs.container.ContainerResponseContext;
+import jakarta.ws.rs.container.ContainerResponseFilter;
+import jakarta.ws.rs.ext.Provider;
+
+
+@Provider
+@Singleton
+public class AccessLogFilter implements ContainerResponseFilter {
+
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/filter/AccessLogFilter.java:41]: expect to remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 38]
```diff
@@ -158,4 +176,244 @@ public String timers() {
...
         return JsonUtil.toJson(reporter.timers());
     }
+
+    @GET
+    @Timed
+    @Produces(APPLICATION_TEXT_WITH_CHARSET)
+    @RolesAllowed({"admin", "$owner= $action=metrics_read"})
+    public String all(@Context GraphManager manager,
+                      @QueryParam("type") String type) {
+
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:186]: expect to remove a blank line

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 39]
```diff
@@ -158,4 +176,244 @@ public String timers() {
...
+                  .append(helpName)
+                  .append(SPACE_STR + UNTYPED + END_LSTR);
+        promMetric.append(helpName)
+                  .append("{version=\"")
+                  .append(ApiVersion.VERSION.toString()).append("\",}")
+                  .append(SPACE_STR + "1.0" + END_LSTR);
+
+        //gauges
+        for (String key : reporter.gauges().keySet()) {
+            final com.codahale.metrics.Gauge<?> gauge
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:236]: can we remove final mark and import com.codahale.metrics.Gauge as possible

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 40]
```diff
@@ -158,4 +176,244 @@ public String timers() {
...
+        result.put("histograms", reporter.histograms());
+        result.put("meters", reporter.meters());
+        result.put("timers", reporter.timers());
+        return JsonUtil.toJson(result);
+    }
+
+    private String baseMetricPrometheusAll() {
+        StringBuilder promMetric = new StringBuilder();
+        ServerReporter reporter = ServerReporter.instance();
+        String helpName = "hugegraph_info";
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:222]: also define a const var

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 41]
```diff
@@ -158,4 +176,244 @@ public String timers() {
...
+                promMetric.append(helpName)
+                          .append(FIVE_MIN_RATE_ATRR)
+                          .append(metric.getFiveMinuteRate() + END_LSTR);
+                promMetric.append(helpName)
+                          .append(FIFT_MIN_RATE_ATRR)
+                          .append(metric.getFifteenMinuteRate() + END_LSTR);
+            }
+        }
+
+        //timer
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:297]: note expected "// xx" style, and it's better to add more details for these comments

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java [hunk 42]
```diff
@@ -158,4 +176,244 @@ public String timers() {
...
+            metricsMap.put(metricsName, entryMetricsMap);
+
+        }
+        return metricsMap;
+    }
+
+    private String statisticsProm(Map<String, Map<String, Object>> metricMap) {
+        StringBuilder promMetric = new StringBuilder();
+
+        //version
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/metrics/MetricsAPI.java:389]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java [hunk 43]
```diff
@@ -24,12 +24,45 @@
...
+    public static final String P95_ATTR = "{name=\"p95\",} ";
+    public static final String P98_ATTR = "{name=\"p98\",} ";
+    public static final String P99_ATTR = "{name=\"p99\",} ";
+    public static final String P999_ATTR = "{name=\"p999\",} ";
+    public static final String MEAN_RATE_ATRR = "{name=\"mean_rate\",} ";
+    public static final String ONE_MIN_RATE_ATRR = "{name=\"m1_rate\",} ";
+    public static final String FIVE_MIN_RATE_ATRR = "{name=\"m5_rate\",} ";
+    public static final String FIFT_MIN_RATE_ATRR = "{name=\"m15_rate\",} ";
+    public static final MetricRegistry REGISTRY =
+            MetricManager.INSTANCE.getRegistry();
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:47]: seems one line is ok

File: hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java (hunk 32 above)
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/metrics/MetricsUtil.java:88]: remove a blank line

File: hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java (hunk 5 above)
Comment (javeme) [reply hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java:192]: not addressed: prefer to keep the origin alignment

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java (hunk 3 above)
Comment (javeme) [reply hugegraph-api/src/main/java/org/apache/hugegraph/api/API.java:99]: not addressed comment

**Review by javeme — APPROVED** [review]
LGTM

File: hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java [hunk 44]
```diff
@@ -234,8 +259,8 @@ public void testMetricsBackend() {
                     String key = (String) e.getKey();
                     value = e.getValue();
                     Assert.assertTrue(String.format(
-                                      "Expect map value for key %s but got %s",
-                                      key, value),
+                                              "Expect map value for key %s but got %s",
```
Comment (JackyYangPassion) [inline hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java]: prefer to keep the origin alignment

Comment (SunnyBoy-WYH) [reply hugegraph-test/src/main/java/org/apache/hugegraph/api/MetricsApiTest.java]: seems done？ if we choose all commit
image

**Review by JackyYangPassion — APPROVED** [review]

**Pull request merged**